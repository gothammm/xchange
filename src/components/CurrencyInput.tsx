import React, { useCallback, KeyboardEvent } from 'react';
import { Input, Typography } from 'antd';
import { WalletType } from '../context/AppContext';
import { getCurrencyDisplayString } from './CurrencyDisplay';
import styled from 'styled-components';
import Text from 'antd/lib/typography/Text';

interface CurrencyInputProps {
  wallet: WalletType;
  value: number;
  maxValue?: number;
  onChange: (value: number) => void;
}

const VALID_FIRST = /^[1-9]{1}$/;
const VALID_NEXT = /^[0-9]{1}$/;
const DELETE_KEY_CODE = 8;

const CurrencyPrefix = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const InputWrapper = styled.div`
  input {
    text-align: right;
    font-size: 23px;
  }
  h3 {
    margin: 0;
  }
`;

const CurrencyInput: React.FC<CurrencyInputProps> = ({
  wallet,
  value,
  onChange,
  maxValue = 100000000,
}) => {
  const valueAbsTrunc = Math.trunc(Math.abs(value));

  if (
    value !== valueAbsTrunc ||
    !Number.isFinite(value) ||
    Number.isNaN(value)
  ) {
    throw new Error(`Invalid Value`);
  }
  const handleKeyDown = useCallback(
    (e: KeyboardEvent<HTMLDivElement>): void => {
      debugger;
      const { key, keyCode } = e;
      if (
        (value === 0 && !VALID_FIRST.test(key)) ||
        (value !== 0 && !VALID_NEXT.test(key) && keyCode !== DELETE_KEY_CODE)
      ) {
        return;
      }
      const valueString = value.toString();
      let nextValue: number;
      if (keyCode !== DELETE_KEY_CODE) {
        const nextValueString: string =
          value === 0 ? key : `${valueString}${key}`;
        nextValue = Number.parseInt(nextValueString, 10);
      } else {
        const nextValueString = valueString.slice(0, -1);
        nextValue =
          nextValueString === '' ? 0 : Number.parseInt(nextValueString, 10);
      }
      if (nextValue > maxValue) {
        return;
      }
      onChange(nextValue);
    },
    [maxValue, onChange, value]
  );
  const valueDisplay = getCurrencyDisplayString(value / 100, wallet.currency);
  return (
    <InputWrapper>
      <Input
        placeholder="Enter value"
        value={valueDisplay}
        onKeyDown={handleKeyDown}
        prefix={
          <CurrencyPrefix>
            <Typography.Title level={3}>{wallet.currency}</Typography.Title>
            <Text type="secondary">
              Balance: {getCurrencyDisplayString(wallet.value, wallet.currency)}
            </Text>
          </CurrencyPrefix>
        }
      />
    </InputWrapper>
  );
};

export default CurrencyInput;
