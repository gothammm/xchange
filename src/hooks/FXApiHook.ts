import { useState } from 'react';
import CurrencyType from '../enums/CurrencyType';

const BASE_URL = `https://api.exchangeratesapi.io`;

interface FXResponse {
  rates: {
    [key: string]: string;
  };
  base: string;
}

interface FXResult {
  from: {
    currency: CurrencyType;
    value: number;
  };
  to: {
    currency: CurrencyType;
    value: number;
  };
  rate: number;
}

export const useFXApiHook = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState();
  const [result, setResult] = useState<FXResult>();

  return {
    error,
    result,
    isLoading,
    fx: (value: number, from: CurrencyType, to: CurrencyType) => {
      setIsLoading(true);
      return fetch(`${BASE_URL}/latest?base=${from}&symbols=${to}`)
        .then(response => response.json())
        .then((fxData: FXResponse) => {
          const conversionRate = +fxData.rates[to];
          setResult({
            from: {
              value,
              currency: from,
            },
            to: {
              value: +(value * conversionRate).toFixed(2),
              currency: to,
            },
            rate: conversionRate,
          });
        })
        .catch(e => setError(e))
        .finally(() => setIsLoading(false));
    },
  };
};
