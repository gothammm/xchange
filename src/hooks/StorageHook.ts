import { useState, useEffect } from 'react';

const STORAGE_KEY = `xchange-data`;

export const useAppStorage = () => {
  const currentData = JSON.parse(
    localStorage.getItem(STORAGE_KEY) || JSON.stringify({})
  );
  const [data, setData] = useState(currentData);

  useEffect(() => {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
  }, [data]);

  return {
    getStorageData: () => data,
    updateStorageData: (data: any) => setData(data),
  };
};
