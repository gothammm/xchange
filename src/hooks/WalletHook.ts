import { useContext } from 'react';
import AppContext, { WalletType } from '../context/AppContext';
import CurrencyType from '../enums/CurrencyType';

export const initialWalletData = [
  {
    currency: CurrencyType.USD,
    value: 0,
    name: 'USD Wallet',
    isPrimary: true,
    id: 0,
  },
  {
    currency: CurrencyType.GBP,
    value: 0,
    name: 'GBP Wallet',
    isPrimary: false,
    id: 1,
  },
  {
    currency: CurrencyType.EUR,
    value: 0,
    name: 'EUR Wallet',
    isPrimary: false,
    id: 3,
  },
];

export const useWallet = () => {
  const appContext = useContext(AppContext);
  return {
    wallets: appContext?.wallets,
    makePrimary: (wallet: WalletType) => {
      const updatedWallets =
        appContext?.wallets?.reduce(
          (acc, item) => [
            ...acc,
            { ...item, isPrimary: wallet.id === item.id },
          ],
          [] as WalletType[]
        ) || [];
      appContext?.update({ ...appContext, wallets: updatedWallets });
    },
    resetWallet: () =>
      appContext?.update({
        ...appContext,
        wallets: initialWalletData,
      }),
    addWallet: (
      name: string,
      currency: CurrencyType,
      initialValue: number = 0
    ) => {
      const newWallet = {
        name,
        currency,
        value: initialValue,
        isPrimary: false,
      } as WalletType;
      appContext?.update({
        ...appContext,
        wallets: (appContext.wallets || []).concat([newWallet]),
      });
    },
  };
};
