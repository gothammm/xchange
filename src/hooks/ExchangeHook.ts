import { WalletType } from '../context/AppContext';
import { useWallet } from './WalletHook';

export const useExchange = () => {
  const {
    addMoneyToWallet,
    deductMoneyFromWallet,
    updateWallet,
    wallets,
    commit,
  } = useWallet();

  return {
    exchange: (
      fromWallet: WalletType,
      amountToDeductFromWallet: number,
      toWallet: WalletType,
      amountToAddToWallet: number
    ) => {
      return Promise.resolve().then(() => {
        if (fromWallet.value - amountToDeductFromWallet < 0) {
          return Promise.reject(
            `Not enough money in wallet with currency ${fromWallet.currency} to exchange`
          );
        }
        const updatedContextWallets = [
          deductMoneyFromWallet(fromWallet, amountToDeductFromWallet, false),
          addMoneyToWallet(toWallet, amountToAddToWallet, false),
        ].reduce((contextWallets, wallet) => {
          return updateWallet(wallet, contextWallets);
        }, (wallets || []) as WalletType[]);

        commit(updatedContextWallets);
        return Promise.resolve();
      });
    },
  };
};
