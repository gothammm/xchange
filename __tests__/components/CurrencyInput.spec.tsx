import React from 'react';
import CurrencyInput from '../../src/components/CurrencyInput';
import { render } from '@testing-library/react';
import { initialWalletData } from '../../src/hooks/WalletHook';

describe(`CurrencyInput component`, () => {
  test(`simple snapshot for different wallet data`, () => {
    initialWalletData.forEach(wallet => {
      const component = render(
        <CurrencyInput value={60} wallet={wallet} onChange={() => {}} />
      );
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
