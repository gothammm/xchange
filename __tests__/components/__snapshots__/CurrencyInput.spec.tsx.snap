// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput component simple snapshot for different wallet data 1`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ dJUKa-D"
  >
    <span
      class="ant-input-affix-wrapper"
    >
      <span
        class="ant-input-prefix"
      >
        <div
          class="sc-AxjAm jsGdgg"
        >
          <h3
            class="ant-typography"
          >
            USD
          </h3>
          <span
            class="ant-typography ant-typography-secondary"
          >
            Balance: $0.00
          </span>
        </div>
      </span>
      <input
        class="ant-input"
        placeholder="Enter value"
        type="text"
        value="$0.60"
      />
    </span>
  </div>
</DocumentFragment>
`;

exports[`CurrencyInput component simple snapshot for different wallet data 2`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ dJUKa-D"
  >
    <span
      class="ant-input-affix-wrapper"
    >
      <span
        class="ant-input-prefix"
      >
        <div
          class="sc-AxjAm jsGdgg"
        >
          <h3
            class="ant-typography"
          >
            GBP
          </h3>
          <span
            class="ant-typography ant-typography-secondary"
          >
            Balance: £0.00
          </span>
        </div>
      </span>
      <input
        class="ant-input"
        placeholder="Enter value"
        type="text"
        value="£0.60"
      />
    </span>
  </div>
</DocumentFragment>
`;

exports[`CurrencyInput component simple snapshot for different wallet data 3`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ dJUKa-D"
  >
    <span
      class="ant-input-affix-wrapper"
    >
      <span
        class="ant-input-prefix"
      >
        <div
          class="sc-AxjAm jsGdgg"
        >
          <h3
            class="ant-typography"
          >
            EUR
          </h3>
          <span
            class="ant-typography ant-typography-secondary"
          >
            Balance: €0.00
          </span>
        </div>
      </span>
      <input
        class="ant-input"
        placeholder="Enter value"
        type="text"
        value="€0.60"
      />
    </span>
  </div>
</DocumentFragment>
`;
